### ISSUE TITLE
Docs: add info about contract compilation with the optimizer

### ISSUE THREAD

**Issue by ebaizel** [trufflesuite/truffle#1284]
This is more of a documentation request around compiling contracts with the optimizer enabled.

**tldr;** It would be very helpful to know when would be the best time to enable the optimizer.

**Background**
I wrote a contract that cost quite a bit to deploy on main net.  I recompiled it with the optimizer turned on, and the cost of the contract was cut in nearly half.  I am trying to understand why this wouldn't be enabled by default.  It seems like there are optimization parameters that tunes it for reducing cost of initial deployment vs tuning it for high frequency usage, so maybe that has something to do with it?

**Add'l info**
From chatting with a couple of folks from the Truffle team, it seems like there could be some concerns / unknowns in regards to the safety of the optimized code, but it wasn't certain.

It was also pointed out that there is a maximum contract size of `0x6000` bytes, which results in an OOG error if the contract exceeds that.  In that case, the only option is to optimize/compress. See [EIP170](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-170.md) for more details.

**Comment by gnidan** [comment]
Thanks for raising this concern. If you'd like, feel free to open a PR with additional documentation, or we will prioritize doing so accordingly.

For background, here is some information about why the optimizer is off by default: https://github.com/trufflesuite/truffle-compile/pull/5

This might be historical and now not necessary. More research warranted.
